[RFC] Add optional third tab character in "listchars"

natbraun opened the pull request:
This PR is a rebase of https://github.com/neovim/neovim/pull/3088

natbraun committed:
Add optional third tab character in "listchars"

* If two characters are specifed, do as before
* If three characters are specified, e.g. <->:
   - if the tab length is 1: display the third character: >
   - if the tab length is 2: display the first and third characters: <>
   - otherwise display the first, repeat the second, display the third: <--->

fmoralesc commented:
It seems like linting errors must be fixed, and a test would be great. (It seems like I closed the previous PR just because I got lazy :p)

natbraun commented:
@fmoralesc Thanks! However, I don't understand the clint errors, I can't relate the lines to the code.
I'd like to write a test, could you please describe the procedure to add one? Thanks!

fmoralesc commented:
@natbraun Sometimes the lint errors come from lines around the changes you made, but I see what you mean, the line offsets are wrong for some reason. For example, the error in line 2779 in screen.c seems to happen in 2783 instead:

~~~ diff
+           c_final = NUL;
+          }
+          else {
             c_extra = 
~~~ 

I suggest you try identifying the kind of errors lint identified in the region surrounding the changes you made.

Now, to add a test I think the best probably is to try to expand some existing test for listchars (`functional/legacy/listchars_spec.lua`), guiding yourself by the existing tests should help you understand how that is supposed to work.

natbraun commented:
@fmoralesc Just added some tests. Thanks for your feedback :)

fmoralesc commented:
@natbraun Great! Now, the only issue is the lint errors. I'll try to take a look, but no promises ;) (so feel free to go ahead)

natbraun commented:
@fmoralesc Thanks a lot for your feedback! Really appreciate it :) Now, I gave a look at the lint errors, but to be honest I don't see where the errors come from... I'll take a closer look now.

natbraun commented:
@fmoralesc OK, I'm starting to see clearer:
1. src/nvim/option.c:3358:  If an else has a brace on one side, it should have it on both: this is true, but it's not been introduced by my code.
2. Same goes for FALSE & lin elength

 I can provide a fix if needed. Thanks.

natbraun committed:
Fix various linting errors

src/nvim/option.c:3358:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
src/nvim/screen.c:2763:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/nvim/screen.c:2763:  Use false instead of FALSE.  [readability/bool] [4]
Braces in various places

fmoralesc commented:
@natbraun As I mentioned earlier, sometimes the linter looks for errors in lines surrounding the changes made. At any rate, the travis build must pass so this can be merged (I think), so if you can, please fix those errors (and any others that might arise).

natbraun commented:
@fmoralesc Looks better now :)

fmoralesc commented on src/nvim/screen.c:
```diff
@@ -2750,7 +2759,9 @@ win_line (
-          n_extra = get_breakindent_win(wp, ml_get_buf(wp->w_buffer, lnum, FALSE));
+          c_final = NUL;
+          n_extra = get_breakindent_win(wp,
+            ml_get_buf(wp->w_buffer, lnum, false));
```
This line should be indented under 'w' in 'wp':
~~~
+          n_extra = get_breakindent_win(wp,
+                                                            ml_get_buf(wp->w_buffer, lnum, false));
~~~
but since that brings the issue of line length, perhaps indenting 2 spaces after `get_breakindent_win` might be better?
~~~
+          n_extra = get_breakindent_win(wp,
+                                                            ml_get_buf(wp->w_buffer, lnum, false));
~~~
Either way, no big deal.

natbraun replied:
OK, fixed using the first solution: it actually fits!

fmoralesc commented on src/nvim/message.c:
```diff
@@ -1399,7 +1400,9 @@ void msg_prt_line(char_u *s, int list)
     if (n_extra > 0) {
       --n_extra;
-      if (c_extra)
+      if (n_extra == 0 && c_final)
```
Please add braces to this conditional, we typically want them even when the body of the clauses is 1 line long, for readability.

natbraun replied:
Done.

fmoralesc commented on src/nvim/option.c:
```diff
@@ -3320,7 +3320,7 @@ static char_u *set_chars_option(char_u **varp)
         if (tab[i].cp != NULL)
           *(tab[i].cp) = (varp == &p_lcs ? NUL : ' ');
       if (varp == &p_lcs)
```
Perhaps add braces to this conditional too?

natbraun replied:
Done, and I put braces around the for loop as well.

fmoralesc commented:
Sure does, and the errors in QB seem to be unrelated. Great work! I made some minor comments.

natbraun commented:
Thanks for your feedback @fmoralesc ! Just pushed an updated & rebased PR.

fmoralesc commented:
Thanks, @natbraun. It looks good to me, but let's wait for a second opinion. @justinmk marked this for 0.2 (this just means that it might land after 0.1 is released, so it shouldn't divert people from working on the more urgent issues), but it might be merged earlier depending on someone reviewing it.

natbraun commented:
@fmoralesc, thanks! Really appreciate it. I'll keep rebasing this branch as the master branch advances, so it's up to date. Now, I'd like to contribute to neovim in general. How can I help? Thanks!

fmoralesc commented:
> How can I help?

@natbraun Just as you did! 

I don't know what's the general status of things, but maybe check on some of the issues for 0.2 and see if something grabs your interest -- or if you have an idea, put it out, research how to implement it and try to do it. There's also the issue of backporting patches from vim, which while tedious at times, gives you some familiarity with the codebase. Any task will give you an appreciation of how the project works, and the "intangibles", like "who's who" and such. People will help, most are friendly -- and if not, at least they are on point ;)

cacplate commented:
Do you think you could also use the test file to test if all works when only two characters? Maybe it doesn't change anything but curious.

justinmk commented:
@k-takata 	Thanks for the notice!

Merged in https://github.com/neovim/neovim/pull/9539
